cmd/gpu_nfdhook/labeler.go

```diff
@@ -171,33 +171,70 @@ func (l *labeler) createCapabilityLabels(cardNum string, numTiles uint64) {
 	}
 	defer file.Close()
 
-	// define strings to search from the file, and the actions to take in order to create labels from those strings (as funcs)
+	gen := ""
+	media := ""
+	graphics := ""
+	// define string prefixes to search from the file, and the actions to take in order to create labels from those strings (as funcs)
 	searchStringActionMap := map[string]func(string){
-		"platform: ": func(platformName string) {
+		"platform:": func(platformName string) {
 			l.labels.addNumericLabel(labelNamespace+"platform_"+platformName+".count", 1)
 			l.labels[labelNamespace+"platform_"+platformName+".tiles"] = strconv.FormatInt(int64(numTiles), 10)
 			l.labels[labelNamespace+"platform_"+platformName+".present"] = "true"
 		},
-		"gen: ": func(genName string) {
-			l.labels[labelNamespace+"platform_gen"] = genName
+		// there's also display block version, but that's not relevant
+		"media version:": func(version string) {
+			l.labels[labelNamespace+"media_version"] = version
+			media = version
+		},
+		"graphics version:": func(version string) {
+			l.labels[labelNamespace+"graphics_version"] = version
+			graphics = version
+		},
+		"gen:": func(version string) {
+			l.labels[labelNamespace+"platform_gen"] = version
+			gen = version
 		},
 	}
 
 	// Finally, read the file, and try to find the matches. Perform actions and reduce the search map size as we proceed. Return at 0 size.
 	scanner := bufio.NewScanner(file)
+scanning:
 	for scanner.Scan() {
-		for searchString, action := range searchStringActionMap {
-			var stringValue string
-			n, _ := fmt.Sscanf(scanner.Text(), searchString+"%s", &stringValue)
-			if n > 0 {
-				action(stringValue)
-				delete(searchStringActionMap, searchString)
-				if len(searchStringActionMap) == 0 {
-					return
-				}
-				break
+		line := scanner.Text()
+		for prefix, action := range searchStringActionMap {
+			if !strings.HasPrefix(line, prefix) {
+				continue
+			}
+			fields := strings.Split(line, ": ")
+			if len(fields) == 2 {
+				action(fields[1])
+			} else {
+				klog.Warningf("invalid '%s' line format: '%s'", file.Name(), line)
+			}
+			delete(searchStringActionMap, prefix)
+			if len(searchStringActionMap) == 0 {
+				break scanning
```

Comment:
Uh. Ok, with the way that you added code below, you indeed can't return with the empty map, as there is some sort of mopping up code and label innovating after the scan actions.

This is just about as much as we can push into a single function, if you go much further the linters start to cry.

Reply:
13 lines would go away if we decide that the "gen" label does not need to be supported any more (instead of being deprecated).  And all of it would disappear if we decide that even *_version labels would not be supported for older kernels that have just "gen" capability (but I think that's a bit too drastic).

Reply:
The capability file originating labels have always been described to be working in a best effort fashion. IMO you don't need to go to lengths trying to maintain any old ones, if they need changing.